src/libsystemd-network/dhcp-static-lease.c

```diff
@@ -0,0 +1,141 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "dhcp-internal.h"
+#include "dhcp-server-internal.h"
+#include "memory-util.h"
+#include "strv.h"
+#include "utf8.h"
+
+#define ADDR_NULL 0
+
+static sd_dhcp_static_lease *current_static_lease = NULL;
+
+static sd_dhcp_static_lease* sd_dhcp_static_lease_free(sd_dhcp_static_lease *i) {
+        if (!i)
+                return NULL;
+
+        free(i->client_id.data);
+
+        return mfree(i);
+}
+
+_destructor_ static void sd_dhcp_static_lease_current_cleanup(void) {
+        sd_dhcp_static_lease_free(current_static_lease);
+        TAKE_PTR(current_static_lease);
+}
+
+static int dhcp_static_lease_new(DHCPClientId *client_id, const struct in_addr *address, sd_dhcp_static_lease **ret) {
+        sd_dhcp_static_lease *p = new(sd_dhcp_static_lease, 1);
+        if (!p)
+                return -ENOMEM;
+
+        assert_return(ret, -EINVAL);
+        assert_return(address, -EINVAL);
+
+        *p = (sd_dhcp_static_lease) {
+                .n_ref = 1,
+                .client_id = *client_id,
+                .address = le32toh(address->s_addr)
+        };
+
+        *ret = TAKE_PTR(p);
+        return 0;
+}
+
+static int dhcp_static_lease_new_default(sd_dhcp_static_lease **ret) {
+        struct in_addr address = { ADDR_NULL };
+        DHCPClientId client_id = { .length = 1, .data = NULL };
+        return dhcp_static_lease_new(&client_id, &address, ret);
+}
+
+int sd_dhcp_static_lease_new_by_mac(const uint8_t *mac_addr, size_t mac_addr_size, const struct in_addr *address, sd_dhcp_static_lease **ret){
+        int r;
+        DHCPClientId c = {
+                .length = ETH_ALEN + 1
+        };
+
+        assert_return(mac_addr, -EINVAL);
+        assert_return(mac_addr_size, -EINVAL);
+
+        c.data = malloc(mac_addr_size + 1);
+        if (!c.data)
+                return log_oom();
+        /* set client id type to 1: Ethernet Link-Layer (RFC 2132) */
+        ((uint8_t*) c.data)[0] = 0x01;
+        memcpy((uint8_t*) c.data + 1, mac_addr, mac_addr_size);
+
+        r = dhcp_static_lease_new(&c, address, ret);
+
+        /* if there was no errors c.data will be freed with the static lease */
+        if (r<0)
+                free(c.data);
+        return r;
+}
+
+int sd_dhcp_static_lease_set_client_id_by_mac(
+        sd_dhcp_static_lease *lease, const uint8_t *mac_addr, size_t mac_addr_size) {
+
+        assert_return(lease, -EINVAL);
+        assert_return(mac_addr, -EINVAL);
+        assert_return(mac_addr_size, -EINVAL);
+
+        if (lease->client_id.data){
+                free(lease->client_id.data);
+        }
+
+        lease->client_id.length = ETH_ALEN + 1;
+        lease->client_id.data = malloc(mac_addr_size + 1);
+        if (!lease->client_id.data) {
+                return log_oom();
+        }
+
```

Comment:
so if the malloc() fails, then you will already have destroyed the old `lease->client_id.data` leaving the pointer pointing to invalid memory, so that the whole object cannot be destructed safely anymore.

Reply:
Wouldn't malloc set pointer to be NULL on failure or am I missing something? So memory will be destroyed and malloc will fail to create a new one, thus leaving pointer `lease->client_id.data` to have value NULL.